### PR TITLE
[Phan] Add PhanUndeclaredTypeReturnType rule to static analysis config

### PR DESCRIPTION
**Pull request by johnsaigle**
### Brief summary of changes

This PR adds the PhanUndeclaredTypeReturnType rule which makes sure that the `return` docs are immediately followed by a type and not an english sentence.

I also made a bunch of small changes to other files because I had a newer version of phan installed at some point which caused other issues to appear.... so this PR will head some of those off by the time we actually update phan.

### CODE REVIEW

**Comment by johnsaigle** [comment]
Many of the errors generated here should be solved by the PRs tracked by #4072

**Comment by driusan** [comment]
What's the status of this? Would it be mergeable if you only added one of the checks instead of both?

**Comment by johnsaigle** [comment]
The status is that it's plodding along. It depends a lot on the typing checks being merged. Both rules are causing issues. It's possible to split this into two PRs so that the changes are more bite-sized but neither could be merged now.

#4153 took care of some of them. https://github.com/aces/Loris/projects/11 will take care of more.

I am still rebasing this script periodically and monitoring the errors.

**Comment by johnsaigle** [comment]
This PR is now only allowing `PhanUndeclaredTypeReturnType` through as opposed to `PhanTypeMismatchReturn`. The latter was generating a lot of false positives that I wasn't able to debug.

**Comment by johnsaigle** [comment]
@driusan This is ready to be reviewed and merged assuming you're fine with the changes.

File: test/unittests/Loris_PHPUnit_Database_TestCase.php [hunk 1]
```diff
@@ -89,6 +89,7 @@ protected function setUp()
      *
      * @return \PHPUnit_Extensions_Database_DB_DefaultDatabaseConnection
      */
+    /* @phan-suppress-next-line PhanUndeclaredTypeReturnType*/
```
Comment (driusan) [inline test/unittests/Loris_PHPUnit_Database_TestCase.php]: Why is this here? It says above it the return type is PHPUnit_Extensions_Database_DB_DefaultDatabaseConnection

File: modules/genomic_browser/php/view_genomic_file.class.inc [hunk 2]
```diff
@@ -146,7 +146,7 @@ class View_Genomic_File extends \NDB_Form
     * @param string      $type  type of the date
     * @param array       $array array containing the date
     *
-    * @return date if exists, if not an empty string
+    * @return int|string Date (Unix int timestamp) if exists, or an empty string
```
Comment (driusan) [inline modules/genomic_browser/php/view_genomic_file.class.inc]: 🤦‍♂️

Comment (driusan) [reply modules/genomic_browser/php/view_genomic_file.class.inc]: (I don't know why GitHub doesn't have a (:picard_facepalming: option))

File: test/unittests/Loris_PHPUnit_Database_TestCase.php (hunk 1 above)
Comment (johnsaigle) [reply test/unittests/Loris_PHPUnit_Database_TestCase.php]: I'm not sure, Phan seemed to have an issue with it at one stage. I'll go back and verify that this is the case.

Comment (johnsaigle) [reply test/unittests/Loris_PHPUnit_Database_TestCase.php]: This issue is fixed in #4274 which is based on this PR. If you merge this one first then this line will be removed and the bug fixed.

File: htdocs/api/v0.0.2/APIBase.php [hunk 3]
```diff
@@ -143,9 +143,9 @@ function handleETag()
     /**
      * Calculate the ETag for this resource
      *
-     * @return string an ETag for this resource
+     * @return ?string an ETag for this resource
      */
-    abstract function calculateETag();
+    abstract function calculateETag(): ?string;
```
Comment (driusan) [inline htdocs/api/v0.0.2/APIBase.php]: I think this should always return a string (the empty string if there's no etag, rather than null), but as @xlecours says this is going to be dead code once the API PR gets merged and the api directory is ignored by phan, so we don't need to do anything about this.

**Comment by johnsaigle** [comment]
@xlecours I'll just revert the changes I made to the API files instead of blocking this.

**Comment by johnsaigle** [comment]
Blocked by #4387

**Comment by driusan** [comment]
@johnsaigle the diff on github says that this literally changed 0 lines of code.. can this be closed or is there something wrong with github?

**Comment by driusan** [comment]
This PR contains no changes according to the diff on GitHub (https://github.com/aces/Loris/pull/4069/files)
